src/Shared/NamedPipeUtil.cs

```diff
@@ -25,5 +35,244 @@ internal static string GetPipeNameOrPath(string pipeName)
+#endif
+            IdentityReference remoteOwner = remoteSecurity.GetOwner(typeof(SecurityIdentifier));
+            if (remoteOwner != identifier)
+            {
+                CommunicationsUtilities.Trace("The remote pipe owner {0} does not match {1}", remoteOwner.Value, identifier.Value);
+                throw new UnauthorizedAccessException();
+            }
+        }
+#endif
+
+        /// <summary>
+        /// Attempts to connect to the specified process.
+        /// </summary>
+        internal static NamedPipeClientStream TryConnectToProcess(int nodeProcessId, int timeout, Handshake? handshake)
+        {
+            string pipeName = GetPipeNameOrPath("MSBuild" + nodeProcessId);
+            return TryConnectToProcess(pipeName, nodeProcessId, timeout, handshake);
+        }
+
+        /// <summary>
+        /// Attempts to connect to the specified process.
+        /// </summary>
+        internal static NamedPipeClientStream TryConnectToProcess(string pipeName, int timeout, Handshake? handshake)
+        {
+            return TryConnectToProcess(pipeName, null, timeout, handshake);
+        }
+
+        private static NamedPipeClientStream TryConnectToProcess(string pipeName, int? nodeProcessId, int timeout, Handshake? handshake)
+        {
+            NamedPipeClientStream nodeStream = new NamedPipeClientStream(".", pipeName, PipeDirection.InOut, PipeOptions.Asynchronous
+#if FEATURE_PIPEOPTIONS_CURRENTUSERONLY
+                                                                         | PipeOptions.CurrentUserOnly
+#endif
+                                                                         );
+           CommunicationsUtilities.Trace("Attempting connect to PID {0} with pipe {1} with timeout {2} ms", nodeProcessId.HasValue ? nodeProcessId.Value.ToString() : pipeName, pipeName, timeout);
+
+            try
+            {
+                nodeStream.Connect(timeout);
+
+#if !FEATURE_PIPEOPTIONS_CURRENTUSERONLY
+                if (NativeMethodsShared.IsWindows && !NativeMethodsShared.IsMono)
+                {
+                    // Verify that the owner of the pipe is us.  This prevents a security hole where a remote node has
+                    // been faked up with ACLs that would let us attach to it.  It could then issue fake build requests back to
+                    // us, potentially causing us to execute builds that do harmful or unexpected things.  The pipe owner can
+                    // only be set to the user's own SID by a normal, unprivileged process.  The conditions where a faked up
+                    // remote node could set the owner to something else would also let it change owners on other objects, so
+                    // this would be a security flaw upstream of us.
+                    ValidateRemotePipeSecurityOnWindows(nodeStream);
+                }
+#endif
+                if (handshake.HasValue)
```

Comment:
Why should the handshake be null? This seems very problematic if it is.

Reply:
This is for RAR node. We don't have to do the check, because the handshake is already part of the name of RAR node pipe. With this we can also save little bit of time (even if it is marginal, it is better than nothing).

Reply:
This doesn't seem ideal from a security perspective. I know the full check doesn't make it bulletproof, but I don't like there being cases when we don't even validate that a node that happens to have the right name is actually a node we want to connect to.

Reply:
Ok, will add the check to the connection.

Reply:
I will wait for this comment to be resolved: https://github.com/dotnet/msbuild/pull/5613#discussion_r478721869. It will be easier if we can upgrade the target framework. Now I would have to modify the callback which is creating the NamePipeServerStream, which is not extra difficult, but still...

Reply:
Added validation